synapse/storage/databases/main/user_directory.py

```diff
@@ -596,75 +609,99 @@ async def update_profile_in_user_dir(
         await self.db_pool.runInteraction(
             "update_profile_in_user_dir",
             self._update_profile_in_user_dir_txn,
-            user_id,
-            display_name,
-            avatar_url,
+            [_UserDirProfile(user_id, display_name, avatar_url)],
         )
 
     def _update_profile_in_user_dir_txn(
         self,
         txn: LoggingTransaction,
-        user_id: str,
-        display_name: Optional[str],
-        avatar_url: Optional[str],
+        profiles: Sequence[_UserDirProfile],
     ) -> None:
-        self.db_pool.simple_upsert_txn(
+        self.db_pool.simple_upsert_many_txn(
             txn,
             table="user_directory",
-            keyvalues={"user_id": user_id},
-            values={"display_name": display_name, "avatar_url": avatar_url},
+            key_names=("user_id",),
+            key_values=[(p.user_id,) for p in profiles],
+            value_names=("display_name", "avatar_url"),
+            value_values=[
+                (
+                    non_null_str_or_none(p.display_name),
+                    non_null_str_or_none(p.avatar_url),
```

Comment:
I don't really see the equivalent of these somewhere in the single-profile version?